cudapoa/include/claraparabricks/genomeworks/cudapoa/batch.hpp

```diff
@@ -56,42 +56,56 @@ struct Entry
-    /// Band-width used in banded alignment
+    int32_t matrix_graph_dimension;
+    /// Maximum horizontal dimension of scoring matrix, which stores part of sequences used in scores matrix computation
+    int32_t matrix_sequence_dimension;
+    /// Band-width used in banded alignment, it also defines minimum band-width in adaptive alignment
     int32_t alignment_band_width;
     /// Maximum number of equences per POA group
     int32_t max_sequences_per_poa;
+    /// Banding mode: full, static, adaptive
+    BandMode band_mode;
 
-    /// constructor- set upper limit parameters based on max_sequence_size
-    BatchSize(int32_t max_seq_sz = 1024, int32_t max_seq_per_poa = 100, int32_t band_width = 256)
+    /// constructor- set upper limit parameters based on max_seq_sz and band_width
+    BatchConfig(int32_t max_seq_sz = 1024, int32_t max_seq_per_poa = 100, int32_t band_width = 256, BandMode banding = BandMode::full_band)
         /// ensure a 4-byte boundary alignment for any allocated buffer
         : max_sequence_size(max_seq_sz)
         , max_consensus_size(2 * max_sequence_size)
-        , max_nodes_per_window(cudautils::align<int32_t, 4>(3 * max_sequence_size))
-        , max_nodes_per_window_banded(cudautils::align<int32_t, 4>(4 * max_sequence_size))
-        , max_matrix_graph_dimension(cudautils::align<int32_t, 4>(max_nodes_per_window))
-        , max_matrix_graph_dimension_banded(cudautils::align<int32_t, 4>(max_nodes_per_window_banded))
-        , max_matrix_sequence_dimension(cudautils::align<int32_t, 4>(max_sequence_size))
-        /// ensure 128-alignment for band_width size
+        /// ensure 128-alignment for band_width size, 128 = CUDAPOA_MIN_BAND_WIDTH
         , alignment_band_width(cudautils::align<int32_t, 128>(band_width))
         , max_sequences_per_poa(max_seq_per_poa)
-
+        , band_mode(banding)
     {
+        if (banding == BandMode::full_band)
+        {
+            max_nodes_per_graph       = cudautils::align<int32_t, 4>(3 * max_sequence_size);
+            matrix_graph_dimension    = cudautils::align<int32_t, 4>(max_nodes_per_graph);
+            matrix_sequence_dimension = cudautils::align<int32_t, 4>(max_sequence_size);
+        }
+        else if (banding == BandMode::static_band)
+        {
+            max_nodes_per_graph    = cudautils::align<int32_t, 4>(4 * max_sequence_size);
+            matrix_graph_dimension = cudautils::align<int32_t, 4>(max_nodes_per_graph);
+            // 8 = CUDAPOA_BANDED_MATRIX_RIGHT_PADDING
+            matrix_sequence_dimension = cudautils::align<int32_t, 4>(alignment_band_width + 8);
+        }
+        else // BandMode::adaptive_band
+        {
+            max_nodes_per_graph    = cudautils::align<int32_t, 4>(4 * max_sequence_size);
+            matrix_graph_dimension = cudautils::align<int32_t, 4>(max_nodes_per_graph);
+            // 8 = CUDAPOA_BANDED_MATRIX_RIGHT_PADDING, *2 is to reserve extra memory for cases with extended band-width
```

Comment:
Having this 8 is not ideal here. I think the reason is that the macro is in one of the `.cu` files and is not accessible from here.

Now it looks like the `BatchConfig` isn't templated, so we should be able to move the implementation into a `batch.cpp` file. and perhaps there we can share a private header file that has the static macros for banding?

Reply:
done, moved to existing `batch.cu` along some other similar changes replacing numbers with static macros.